cabal-validate/main/Main.hs

```diff
@@ -0,0 +1,958 @@
+
+      tastyArgs =
+        "--hide-successes"
+          : case opts.tastyPattern of
+            Just tastyPattern -> ["--pattern", tastyPattern]
+            Nothing -> []
+
+  when opts.listSteps $ do
+    -- TODO: This should probably list _all_ available steps, not just the selected ones!
+    tPutStrLn "Targets:"
+    forM_ targets $ \target -> do
+      tPutStrLn $ "  " <> target
+    tPutStrLn "Steps:"
+    forM_ steps $ \step -> do
+      tPutStrLn $ "  " <> displayStep step
+    exitSuccess
+
+  startTime <- getAbsoluteTime
+  jobs <- maybe getNumCapabilities pure opts.jobs
+  cwd <- getCurrentDirectory
+  compiler <- makeCompiler opts.compiler
+
+  pure
+    ResolvedOpts
+      { verbose = opts.verbose
+      , jobs = jobs
+      , cwd = cwd
+      , startTime = startTime
+      , compiler = compiler
+      , extraCompilers = opts.extraCompilers
+      , cabal = opts.cabal
+      , archPath = archPath
+      , projectFile = projectFile
+      , hackageTests = opts.hackageTests
+      , tastyArgs = tastyArgs
+      , targets = targets
+      , steps = steps
+      }
+
+-- | Command-line options.
+data Opts = Opts
+  { verbose :: Bool
+  , jobs :: Maybe Int
+  , compiler :: FilePath
+  , cabal :: FilePath
+  , extraCompilers :: [FilePath]
+  , tastyPattern :: Maybe Text
+  , doctest :: Bool
+  , steps :: [Step]
+  , listSteps :: Bool
+  , libOnly :: Bool
+  , runLibTests :: Bool
+  , runCliTests :: Bool
+  , runLibSuite :: Bool
+  , runCliSuite :: Bool
+  , solverBenchmarks :: Bool
+  , hackageTests :: HackageTests
+  }
+  deriving (Show)
+
+optsParser :: Parser Opts
+optsParser =
+  Opts
+    <$> ( flag'
+            True
+            ( short 'v'
+                <> long "verbose"
+                <> help "Always display build and test output"
+            )
+            <|> flag
+              False
+              False
+              ( short 'q'
+                  <> long "quiet"
+                  <> help "Silence build and test output"
+              )
+        )
+    <*> option
+      (Just <$> auto)
+      ( short 'j'
+          <> long "jobs"
+          <> help "Passed to `cabal build --jobs`"
+          <> value Nothing
+      )
+    <*> strOption
+      ( short 'w'
+          <> long "with-compiler"
+          -- TODO: For tests? Builds?
+          <> help "Use the given compiler instead of `ghc`"
+          <> value "ghc"
+      )
+    <*> strOption
+      ( long "with-cabal"
+          -- TODO: For builds?
+          <> help "Use the given `cabal-install`"
+          <> value "cabal"
+      )
+    <*> many
+      ( strOption
+          ( long "extra-hc"
+              <> help "Extra compilers to run the test suites with"
+          )
+      )
+    <*> option
+      (Just <$> Opt.str)
+      ( short 'p'
+          <> long "pattern"
+          <> help "Pattern to filter tests by"
+          <> value Nothing
+      )
+    <*> boolOption
+      False
+      "doctest"
+      ( -- TODO: Which libraries?
```

Comment:
In general, this distinguishes `Cabal` (with `Cabal-syntax` being considered a sub-component) from `cabal-install` (aka "cli"). This also applies to the later ones.